src/common/dav_v2/heap.c

```diff
@@ -600,6 +772,85 @@ heap_mbrt_mb_reclaim_garbage(struct palloc_heap *heap, uint32_t zid)
 	return 0;
 }
 
+void
+heap_soemb_active_iter_init(struct palloc_heap *heap)
+{
+	heap->rt->smbrt.cur_idx = 0;
+}
+
+uint32_t
+heap_soemb_active_get(struct palloc_heap *heap)
+{
+	struct soemb_rt *smbrt = &heap->rt->smbrt;
+	struct mbrt     *mb    = NULL;
+
+	if (heap->rt->nzones_e == 0)
+		return 0;
+
+	if (smbrt->cur_idx > smbrt->fur_idx)
+		smbrt->fur_idx = smbrt->cur_idx;
+
+	if (smbrt->cur_idx < SOEMB_ACTIVE_CNT) {
+		mb = smbrt->svec[smbrt->cur_idx];
+		smbrt->cur_idx++;
+	}
+
+	if (mb)
+		return mb->mb_id;
+
+	return 0;
+}
+
+static int
+heap_create_soe_mb(struct palloc_heap *heap, uint32_t *mb_id);
+
+void
+heap_soemb_reserve(struct palloc_heap *heap)
+{
+	int              i, ret;
+	uint32_t         mb_id;
+	struct mbrt     *mb;
+	struct soemb_rt *smbrt = &heap->rt->smbrt;
+
+	if (heap->rt->nzones_e == 0)
+		return;
+
+	if (smbrt->fur_idx > 1) {
+		mb = smbrt->svec[0];
+		if (mb)
+			mbrt_qbs_insertmb(&smbrt->qbs, mb);
+
+		for (i = 1; i < SOEMB_ACTIVE_CNT; i++) {
+			smbrt->svec[i - 1] = smbrt->svec[i];
+		}
+
+		smbrt->svec[SOEMB_ACTIVE_CNT - 1] = NULL;
+		smbrt->fur_idx                    = 0;
+	}
+
+	for (i = 0; i < SOEMB_ACTIVE_CNT; i++) {
+		if (smbrt->svec[i] != NULL)
+			continue;
+		mb = mbrt_qbs_getmb(&smbrt->qbs, 0);
+		if (mb) {
+			smbrt->svec[i] = mb;
+			break;
+		}
+		ret = heap_create_soe_mb(heap, &mb_id);
+		if (ret == 0) {
+			smbrt->svec[i] = heap_mbrt_get_mb(heap, mb_id);
+			break;
+		}
+		mb = mbrt_qbs_getmb(&smbrt->qbs, 1);
+		if (mb) {
+			smbrt->svec[i] = mb;
+			break;
+		}
+		break;
+	}
```

Comment:
Above loop is to create ensure there is always 1 available SOE bucket or SOEMB_ACTIVE_CNT buckets? (The loop breaks if any bucket setup successfully).

I think the goal here is to ensure enough free space being pinned in memory for the potential spilling over happened in next transaction, right? So we'd replace any unqualified bucket (in 'svec') with qualified one to satisfy the space requirement. It looks to me it's too late to remove the unqualified bucket from 'svec' in heap_recycle_soembs().